[php] Fix PDOClass and complete it.

TheoSabattie opened the pull request:
- Add attributes to PDOClass
- Add native macro
- Uncomment PDOClass constructor
- Add overload instead of optional param 
  (when function are called with optional param. => Param = null, php functions check with isset function and that function return true if null is assigned to variable)
- Fix Exception method return type
- Fix Exception constructor
- Complete Exception class
- Add RuntimeException
- Add PDOException

nadako commented:
@mockey Could you review this?

TheoSabattie commented:
I will add some documentations on the week.

mockey commented:
I don't know. I guess it could be added like this, but I have a couple of suggestions:

1. adding class constants with `untyped __php__` in  a getter is really awkward. We should add a `@:phpConst` metadata for this. I can try this next week.
2. Maybe add the mysql driver constants as well: http://php.net/manual/en/ref.pdo-mysql.php#pdo-mysql.constants. I needed some of these recently.
3. rename and separate the classes:
 - `php.db.PDO` -> `php.PDOConnection`
 - `php.db.PDOClass` -> `php.PDO`
 - `php.db.PDOStatement`-> `php.PDOStatement`

About exceptions:
- Does the `Exception` extern even work without `@:native("\\Exception")`? I know there is some special hard-coded handling of `try...catch` in the generator, but `new php.Exception(...)` gives you `new php_Exception(...)` I think as of now, which is not quite right.
- Is `RuntimeException` really needed?
- `PDOException` should be in its own class file.

TheoSabattie commented:
Thanks for review Mockey, 

I added mysql driver constants.
I rename and separate the classes. 
PDOClass is now PDO.
There is one PDO Class in php.PDO and one in PDOConnection. 
It does not matter?

I remove @;native (I added it because PDOClass needed to be PDO). So Exception don't need @;native macro.

I added RuntimeException to respect inheritance of php class.

I noticed wrong something:
`extern class Exception {
	public function new(message : String = "", code : Int = 0, previous:Exception = null) : Void;
// ...`

TheoSabattie commented:
Thanks for review mockey

> Maybe add the mysql driver constants as well: http://php.net/manual/en/ref.pdo-mysql.php#pdo-mysql.constants. I needed some of these recently.

It is done.

> rename and separate the classes:
> php.db.PDO -> php.PDOConnection
> php.db.PDOClass -> php.PDO
> php.db.PDOStatement-> php.PDOStatement

If I renamed PDOClass to PDO I could remove @:native.
But I can not rename php.db.PDO to PDOConnection because there is already private class PDOConnection.
What name should I use for class and file?
We can keep 2 PDO Class in different files. But if we do that there is some issue with unit test for php.
I try to fix it but I did not success for the moment.

> Does the Exception extern even work without @:native("\\Exception")?
Yes.

I noticed wrong something:
Thanks for review mockey

> Maybe add the mysql driver constants as well: http://php.net/manual/en/ref.pdo-mysql.php#pdo-mysql.constants. I needed some of these recently.

It is done.

> rename and separate the classes:
> php.db.PDO -> php.PDOConnection
> php.db.PDOClass -> php.PDO
> php.db.PDOStatement-> php.PDOStatement

If I renamed PDOClass to PDO I could remove @:native.
But I can not rename php.db.PDO to PDOConnection because there is already private class PDOConnection.
What name should I use for class and file?
We can keep 2 PDO Class in different files. But if we do that there is some issue with unit test for php.
I try to fix it but I did not success for the moment.

> Does the Exception extern even work without @:native("\\Exception")?
Yes.

I noticed wrong something:
```
extern class Exception {
	public function new(message : String = "", code : Int = 0, previous:Exception = null) : Void;
// ...
// other class
new Exception();

// output:
new Exception(null, null, null)
```

Default parameters don't work.

> Is RuntimeException really needed?

I respected the inheritance of PDOException.

TheoSabattie commented:
@mockey 

`package sys.db;

import sys.db.Connection;

@:coreApi class Mysql {

	public static function connect( params : {
		host : String,
		?port : Int,
		user : String,
		pass : String,
		?socket : String,
		database : String
	} ) : sys.db.Connection {
		var dsn = "mysql:";
        
		if (params.socket !=null)
			dsn+="unix_socket="+params.socket+";";
		else{
			dsn+="host="+params.host+";";
			if (params.port!=null)
				dsn+='port='+params.port+";";
		}
		dsn+="dbname="+params.database;
		return php.PDOConnection.PDO.open(dsn,params.user,params.pass);
	}

}
`

Don't work.

`package php;

import php.NativeArray;

import sys.Connection;
import sys.ResultSet;

import php.Lib;

class PDO
{
	public static function open(dsn : String, ?user : String, ?password : String, ?options : Dynamic) : Connection {
		return new PDOConnection(dsn, user, password, options);
	}
}

private class PDOConnection implements Connection {

	var pdo : PDOClass;
	var dbname : String;

	public function new(dsn : String, ?user : String, ?password : String, ?options : Dynamic) {
		if(null == options)
			pdo = untyped __call__("new PDO", dsn, user, password);
		else
		{
/...`

`package php;

import php.NativeArray;

import php.Lib;

extern class PDO
{
//...`

Should I put open static function on PDOConnection and remove PDO in php.PDOConnection.hx ?

mockey commented:
Sorry for the confusion. I don't have time for this at the moment. Renaming like this doesn't make sense. I have to think about it.
Maybe just keep the old names/classes for now if you want to merge this.

TheoSabattie commented:
I am trying to fix wrong call when there is optional parameter on php-generator. c.f. [https://github.com/HaxeFoundation/haxe/issues/5318](url)
I think that is the better way instead of overloading all functions.

I have some difficulty to understand the logic of the generator for the moment. There is a lot of recursivity.

If you know to fix that quickly, please, do it.

P.S. What is the better way to comment overloaded functions ?

mockey commented:
> If you know how to fix that quickly, please, do it.

Sorry, I don't know how to fix that quickly, but I'll try to have look later this week.

> What is the better way to comment overloaded functions ?

What do you mean by that?

TheoSabattie commented:
> Sorry, I don't know how to fix that quickly, but I'll try to have look later this week.

Ok' thanks, I will try too but I think you will find solution before me.

> What do you mean by that?

```
        @:overload(function(statement : String, mode : Int):PDOStatement { })
	@:overload(function(statement : String, mode : Int, colno:Int):PDOStatement { })
	@:overload(function(statement : String, mode : Int, object:Dynamic):PDOStatement { })
	@:overload(function(statement : String, mode : Int, classname:String, ctorargs:NativeArray):PDOStatement { })
	public function query(statement : String) : PDOStatement;
```

How can I document that function?

nadako commented:
> How can I document that function?

See #5201 about that

TheoSabattie commented:
I didn't succeed to fix wrong call with default parameters.
You can take this pull request to fix issues and update files when you will make @:phpConst and will fix wrong call.

TheoSabattie commented:
@mockey : I am seeing you added @;phpConst -> https://github.com/HaxeFoundation/haxe/pull/5071
But I don't understand how can I use that for Class like PDO.
Can you give me a sample, please?

mockey commented:
@TheoSabattie :
I added @:phpConstants meta some time ago but it was meant for a whole haxe enum abstract of constants, not for php class constants like in PDO. I would be possible to put all PDO constants in a separate class with this, but that's not really nice and probably not worth the hassle.
Might be better to add a proper PDO implementation to the new php target that @RealyUniqueName is working on. I think he got php constant handling covered already.

RealyUniqueName commented:
Accessing third-party PHP class constants is not implemented in new generator yet. Here is an issue: https://github.com/RealyUniqueName/haxe/issues/60

Simn commented:
@RealyUniqueName: What should we do with this?

RealyUniqueName commented:
I'll be able to figure this out on the next week.

nadako commented on std/php/Exception.hx:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C)2005-2016 Haxe Foundation
+ * Copyright (C)2005-2012 Haxe Foundation
```
really?:)

RealyUniqueName commented:
@TheoSabattie Does this PR really need to change `haxelib` and `libs` submodules?

TheoSabattie commented:
@RealyUniqueName Absolutely not. It is first time I use git with submodule. I don't understand what I did to update it. 

How can I remove my changes on them?

I think I don't handle correctly the pull request.

With your work with php constant -> `@:phpClassConst`, Should I update all?

RealyUniqueName commented:
> How can I remove my changes on them?
I'm not fluent with submodules too. I think, you can merge dev to your branch and resolve submodules conflicts to "theirs" versions.

> With your work with php constant -> @:phpClassConst, Should I update all?
That meta is not supported by old php generator, so there is no need to rework this pull request.

TheoSabattie commented:
@RealyUniqueName I pulled each submodule. haxelib_src is still changed.
But there is no more problem with CI.

RealyUniqueName commented on std/php/Exception.hx:
```diff
@@ -22,19 +22,23 @@
+	public function getCode() : Dynamic;             // code of the exception
+	public function getFile() : String;          // source filename
+	public function getLine() : Int;             // source line
+	public function getTrace() : Array<String>;  // an array of the backtrace()
```
It should be `php.NativeArray` instead of `Array`

TheoSabattie replied:
fix in e9e19af

RealyUniqueName commented on std/php/PDO.hx:
```diff
@@ -19,85 +19,36 @@
  * DEALINGS IN THE SOFTWARE.
  */
-package php.db;
+package php;
```
Is it really needed to move this module to another package?
Looks like not much of a benefit comparing to introducing a breaking change.

TheoSabattie replied:
fix in 52d8363

RealyUniqueName commented on std/php/PDOClass.hx:
```diff
@@ -0,0 +1,197 @@
+import php.Lib;
+
+@:native("PDO")
+extern class PDOClass
```
Perhaps it's beeter to keep this class in `PDO` module, to avoid unnecessary breaking changes.

TheoSabattie replied:
Does that mean to put this class in PDO.hx file?

TheoSabattie replied:
fix in 80f4ad2

RealyUniqueName commented on std/php/PDOClass.hx:
```diff
@@ -0,0 +1,197 @@
+	
+
+    
+    
```
Too many empty lines

TheoSabattie replied:
fix in c3de31e

RealyUniqueName commented on std/php/PDOClass.hx:
```diff
@@ -0,0 +1,197 @@
+    public static var MYSQL_ATTR_MULTI_STATEMENTS(get, never):Int;
+	private static inline function get_MYSQL_ATTR_MULTI_STATEMENTS():Int return untyped __php__("PDO::MYSQL_ATTR_MULTI_STATEMENTS");
+
+    
```
empty lines

TheoSabattie replied:
fix in c3de31e

RealyUniqueName commented on std/php/PDOClass.hx:
```diff
@@ -0,0 +1,197 @@
+	public function quote(string : String, parameter_type : Int) : String;
+	public function rollBack() : Bool;
+	public function setAttribute(attribute : Int, value : Dynamic) : Bool;
+}
```
Missing `inTransaction` method  http://php.net/manual/en/pdo.intransaction.php

TheoSabattie replied:
fix in c3de31e

RealyUniqueName commented on std/php/PDOStatement.hx:
```diff
@@ -0,0 +1,79 @@
+
+import php.Lib;
+
+extern class PDOStatement
```
Perhaps it's beeter to keep this class in PDO module, to avoid unnecessary breaking changes.

TheoSabattie replied:
fix in 80f4ad2

RealyUniqueName commented on std/php/PDO.hx:
```diff
@@ -0,0 +1,584 @@
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+package php.db;
```
Forgot to move this file to `php/db` directory.

TheoSabattie replied:
fix in 1a1c79e

RealyUniqueName commented on std/php/_std/sys/db/Sqlite.hx:
```diff
@@ -24,7 +24,7 @@ package sys.db;
 
 	public static function open( file : String ) : Connection {
-		return php.db.PDO.open("sqlite:"+file);
+		return php.PDO.open("sqlite:"+file);
```
Invalid package.

TheoSabattie replied:
Fix in ee0c665

RealyUniqueName commented:
Thanks for your work! I think after these last few changes this PR can be merged.

TheoSabattie commented:
Do I forget something @RealyUniqueName ?

RealyUniqueName commented:
Not you. It's I forgot about this PR :)
Sorry.
